### PR TITLE
Added support for pattern matching syntax to smart variable completion provider

### LINKED ISSUE #23497: Smart variable naming should support pattern matching syntax
**Version Used**: 
VS 2017 15.4

**Steps to Reproduce**:

```csharp
using System.Threading;

public class C
{
    public static bool Foo()
    {
        object obj = null;
        return obj is CancellationToken $$
    }
}
```
1. Put the cursor to $$
2. Try to invoke completion

**Expected Behavior**:
Completion should provide variable names

**Actual Behavior**:
Nothing happen

### PR DESCRIPTION
**Pull request by zaytsev-victor**
### Customer scenario

The pattern matching syntax is used with an explicit type name, but code completion does not assist the user in typing the name of the new variable.

### Bugs this fixes

Fixes #23497

### Workarounds, if any

Type the variable name.

### Risk

Low. The new code has many tests and only applies to pattern matching cases.

### Performance impact

Negligible. The algorithm used for pattern matching syntax is the same as the one used for other declaration locations.

### Is this a regression from a previous update?

No.

### Root cause analysis

Lack of testing for this case of declaring variables. Tests are now in place.

### How was the bug found?

Reported by external user.

### Test documentation updated?

No.

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 1]
```diff
@@ -296,6 +292,41 @@ internal struct NameDeclarationInfo
...
+                int position, CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                result = default;
+                if (token.Parent.IsParentKind(SyntaxKind.IsExpression))
+                {
+                    result = IsLastTokenOfType<BinaryExpressionSyntax>(
+                        token, semanticModel,
+                        b => b.Right,
+                        _ => default(SyntaxTokenList),
+                        _ => ImmutableArray.Create(SymbolKind.Parameter),
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs]: :bulb: Consider caching this `ImmutableArray<SymbolKind>` instance as a static field.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 2]
```diff
@@ -296,6 +292,41 @@ internal struct NameDeclarationInfo
...
+            private static bool IsPatternMatching(SyntaxToken token, SemanticModel semanticModel,
+                int position, CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                result = default;
+                if (token.Parent.IsParentKind(SyntaxKind.IsExpression))
+                {
+                    result = IsLastTokenOfType<BinaryExpressionSyntax>(
+                        token, semanticModel,
+                        b => b.Right,
+                        _ => default(SyntaxTokenList),
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs]: might be able to simplify this to ```_ => default```

**Comment by chborl** [comment]
@dotnet-bot retest ubuntu_14_debug_prtest please

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 3]
```diff
@@ -1,26 +1,23 @@
...
 using Microsoft.CodeAnalysis.Shared.Extensions;
-using Microsoft.CodeAnalysis.Shared.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Completion.Providers
 {
     internal partial class DeclarationNameCompletionProvider
     {
         internal struct NameDeclarationInfo
         {
+            private static readonly ImmutableArray<SymbolKind> parameterSyntaxKind = ImmutableArray.Create(SymbolKind.Parameter);
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs]: ❕ This should be `ParameterSyntaxKind` or `s_parameterSyntaxKind` (I'll get this fixed)

**Comment by sharwell** [comment]
@Pilchie for ask mode

**Comment by Pilchie** [comment]
Approved for 15.6 Preview 4.

**Pull request merged**